### PR TITLE
issue-03: Implementation Stack container

### PR DESCRIPTION
**Pull request by akisemenovych**

### CODE REVIEW

File: src/core/lifo/main/Stack.kt [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    }
+
+    /**
+     * Stacks iterator
+     *
+     * @author Volodymyr Semenovych (@akisemenovych).
+     * @since 0.1
+     * @return Stacks iterator
+     */
+    override fun iterator(): MutableIterator<E> {
```
Comment (Fly-Style) [inline src/core/lifo/main/Stack.kt]: Stack does not need iterator for it.

File: src/core/lifo/main/Stack.kt [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+     * @author Volodymyr Semenovych (@akisemenovych).
+     * @since 0.1
+     */
+    override fun clear() {
+        this.stackHead = null
+        stackSize = 0
+    }
+
+    /**
+     * Removes elements collection from Stack
```
Comment (Fly-Style) [inline src/core/lifo/main/Stack.kt]: Retain is not removing elements from collection.

File: src/core/lifo/main/Stack.kt [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+     *
+     * @author Volodymyr Semenovych (@akisemenovych).
+     * @since 0.1
+     * @param elements
+     * @return flag of action status
+     */
+    override fun retainAll(elements: Collection<E>): Boolean {
```
Comment (Fly-Style) [inline src/core/lifo/main/Stack.kt]: You doesn`t need retain operation for stack.

File: src/core/lifo/test/StackTest.kt [hunk 4]
```diff
@@ -0,0 +1,83 @@
+import org.junit.Assert
+import org.junit.Test
+
+/**
+ * Created by akisemenovych on 02.05.2017.
+ * @See Stack
+ */
+
+class StackTest {
+    class ClassicListTest {
```
Comment (Fly-Style) [inline src/core/lifo/test/StackTest.kt]: Copy-paste, really?

File: src/core/lifo/main/Stack.kt [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    /**
+     *
+     * Remove elements collection from Stack
+     *
+     * @author Volodymyr Semenovych (@akisemenovych).
+     * @since 0.1
+     * @param elements collection that need to be removed from Stack
+     * @return flag of action status
+    */
+    override fun removeAll(elements: Collection<E>): Boolean {
```
Comment (Fly-Style) [inline src/core/lifo/main/Stack.kt]: Change `removeAll(elements: Collection<E>)` to `removeN(count : Int)`, where you will peek only first N elements from stack if they will exist.

File: src/core/lifo/main/Stack.kt [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+        override fun hashCode(): Int {
+            var result = el?.hashCode() ?: 0
+            result = 31 * result + (next.hashCode() ?: 0)
+            return result
+        }
+    }
+
+    /**
+     * Iterator class for Stack
```
Comment (Fly-Style) [inline src/core/lifo/main/Stack.kt]: Are you really think, that collection with O(1) access to element  and O(1) putting element to container needs iterator?

File: easylib.iml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Fly-Style) [inline easylib.iml]: Also, your commit message doesn`t meets the requirements to commit messages in our job.

File: src/core/lifo/main/Stack.kt (hunk 6 above)
Comment (akisemenovych) [inline src/core/lifo/main/Stack.kt]: That was your thought by asking to implement 
`: MutableCollection<E>` that have abstract iterator

File: src/core/lifo/main/Stack.kt [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+ * removed.
+ *
+ * @author Volodymyr Semenovych (@akisemenovych).
+ * @since 0.1
+ * @param E the type of elements which are consists in this data structure.
+ * @see Node .
+ * @see StackIterator .
+ */
+
+class Stack<E> (private var stackHead : Node<E>? = null): MutableCollection<E> {
```
Comment (akisemenovych) [inline src/core/lifo/main/Stack.kt]: Nice timing to tell it =)

File: src/core/lifo/main/Stack.kt [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+
+    /**
+     * Remove element from Stack
+     *
+     * @author Volodymyr Semenovych (@akisemenovych).
+     * @since 0.1
+     * @param elements collection that need to be added into Stack
+     * @return flag of action status
+     */
+    override fun remove(element: E): Boolean {
```
Comment (akisemenovych) [inline src/core/lifo/main/Stack.kt]: It is implemented method.

File: src/core/lifo/main/Stack.kt (hunk 5 above)
Comment (akisemenovych) [inline src/core/lifo/main/Stack.kt]: Once more, implemented method that required by ` : MutableCollection<E>`

File: src/core/lifo/test/StackTest.kt (hunk 4 above)
Comment (akisemenovych) [inline src/core/lifo/test/StackTest.kt]: Is there different requirements or implementations for different collections? I thougth same requirements, different implementations.

File: src/core/lifo/main/Stack.kt (hunk 3 above)
Comment (akisemenovych) [inline src/core/lifo/main/Stack.kt]: Look at `override` statement

File: src/core/lifo/main/Stack.kt (hunk 8 above)
Comment (Fly-Style) [inline src/core/lifo/main/Stack.kt]: Yes, my bad, sorry.

File: src/core/lifo/main/Stack.kt (hunk 6 above)
Comment (Fly-Style) [inline src/core/lifo/main/Stack.kt]: Yes, it was my mistake, but you are implementing this data structure following the specification. Why not to point me to my mistake?

File: src/core/lifo/test/StackTest.kt (hunk 4 above)
Comment (Fly-Style) [inline src/core/lifo/test/StackTest.kt]: Of course, different. Stack is container with O(1) access or peeking element and O(1) putting element to container. That`s all.

Comment (akisemenovych) [inline src/core/lifo/test/StackTest.kt]: O(1) is time for accessing first element in the Stack (Stack Head). O(n) is cumulative time for n'th element. That test shows all operations above collection, so i cant see any reason to change it. Although we can discuss modifications :)

File: src/core/lifo/main/Stack.kt (hunk 6 above)
Comment (akisemenovych) [inline src/core/lifo/main/Stack.kt]: We told about removing `Surielazible()` interface, so i had thought that if `: MutableCollection<E>` is out of our target here we'd discuss it as well.

File: src/core/lifo/main/Stack.kt (hunk 2 above)
Comment (akisemenovych) [inline src/core/lifo/main/Stack.kt]: yes, thanks for pointing me on this typo mistake, i'll fix it

File: src/core/lifo/main/Stack.kt [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+         * @author Volodymyr Semenovych (@akisemenovych).
+         * @since 0.1
+         */
+        override fun remove() {
+            this.current!!.remove()
+        }
+    }
+
+    fun head(): Node<E>? = this.stackHead
+    fun next(): Node<E>? = this.stackHead!!.next()
```
Comment (akisemenovych) [inline src/core/lifo/main/Stack.kt]: Good advise, i'll change that

File: src/core/lifo/main/Stack.kt [hunk 11]
```diff
@@ -133,51 +79,66 @@ class Stack<E> (private var stackHead : Node<E>? = null): MutableCollection<E> {
...
      * @author Volodymyr Semenovych (@akisemenovych).
      * @since 0.1
-     * @param element that needs to be added into Stack
+     * @param element that needs to be pushed into Stack
      * @return flag of action status
      */
-    override fun add(element : E): Boolean {
+    fun push(element : E): Boolean {
         var current : Node<E>? = null
         if (current != null) {
```
Comment (toniukan) [inline src/core/lifo/main/Stack.kt]: Seems to be weird according to the previous line (assigning to null and checking if not null)

File: src/core/lifo/test/StackTest.kt [hunk 12]
```diff
@@ -7,16 +7,13 @@ import org.junit.Test
  */
 
 class StackTest {
```
Comment (toniukan) [inline src/core/lifo/test/StackTest.kt]: You're not even checking if pop() returns a correct value. Add more tests.

File: src/core/lifo/main/Stack.kt (hunk 11 above)
Comment (akisemenovych) [inline src/core/lifo/main/Stack.kt]: There should be checking "element" on null, already fixed for next commit

**Pull request merged**